plan/heuristic_planner.go

```diff
@@ -126,7 +126,7 @@ func (p *heuristicPlanner) Plan(ctx context.Context, inputPlan *Spec) (*Spec, er
 					nodeStack = append(nodeStack, newNode.Predecessors()[i-1])
 				}
 
-				visited[newNode] = struct{}{}
+				visited[node] = struct{}{}
```

Comment:
I wasn't sure if I should be adding both new and old nodes to the visited map. Thoughts?

Reply:
As background, this traversal starts from the sinks (nodes with no successors) and moves towards the sources, in a depth-first fashion. I still find the term "roots" to be really confusing in this context.

I'm not sure why that order was chosen (I think it was me that chose it) but it seems like we would want to traverse in the opposite direction, from the sources to the sinks, with the additional condition that we don't visit any node until we've visited all its predecessors.

That might be a big change for a bug fix. For this PR, you could keep the traversal direction the same, but only add a node to the nodestack if the current node is the last remaining successor? If that makes sense. I guess in that case you would want to add `newNode` to the `visited` map.

If I understand the problem that's occurring correctly, I think it would solve the issue and also avoid the confusion of any particular node (or new replacement of an old node) appearing more than once in the node stack.